executor/src/witgen/data_structures/identity.rs

```diff
@@ -0,0 +1,430 @@
+
+use powdr_ast::{
+    analyzed::{
+        AlgebraicExpression, AlgebraicUnaryOperator, ConnectIdentity, Identity as AnalyzedIdentity,
+        LookupIdentity, PermutationIdentity, PhantomLookupIdentity, PhantomPermutationIdentity,
+        PolynomialIdentity, SelectedExpressions,
+    },
+    parsed::visitor::Children,
+};
+use powdr_number::FieldElement;
+
+#[derive(Clone, Eq, PartialEq, Hash, Debug)]
+pub struct BusSend<T> {
+    /// The ID is globally unique among identities.
+    pub id: u64,
+    /// The ID of the bus interaction, used to find a matching receive.
+    pub interaction_id: AlgebraicExpression<T>,
+    /// The tuple sent to the bus, with a selector.
+    pub selected_tuple: SelectedExpressions<T>,
+}
+
+#[derive(Clone, Eq, PartialEq, Hash, Debug)]
+pub struct BusReceive<T> {
+    /// The ID is globally unique among identities.
+    pub id: u64,
+    /// The ID of the bus interaction, to be matched with a send.
+    pub interaction_id: T,
+    /// The tuple sent to the bus, with a selector.
+    /// The selector is an expression that might or might not reference
+    /// the multiplicity column, but it should always evaluate to a
+    /// binary value. If it is zero, the multiplicity must be zero as well.
+    /// For example, it could also be binary fixed column, indicating
+    /// where the multiplicity can be non-zero.
+    pub selected_tuple: SelectedExpressions<T>,
+    /// The multiplicity with which the tuple is sent to the bus.
+    /// None if this interaction comes from the RHS of a native lookup,
+    /// as native lookup arguments don't require the multiplicity to be
+    /// committed.
+    /// Note that this is the absolute multiplicity, i.e., for bus receives,
+    /// the negation has been removed.
+    pub multiplicity: Option<AlgebraicExpression<T>>,
+}
+
+impl<T: FieldElement> BusSend<T> {
+    /// Tries to find the matching receive, which should succeed iff. the send has
+    /// a static interaction ID.
+    pub fn try_match_static<'a>(
+        &self,
+        receives: &'a BTreeMap<T, BusReceive<T>>,
+    ) -> Option<&'a BusReceive<T>> {
+        Some(&receives[&self.interaction_id()?])
+    }
+
+    /// Returns the interaction ID if it is a number.
+    /// Sends and receives can be matched by matching the interaction ID.
+    pub fn interaction_id(&self) -> Option<T> {
+        match &self.interaction_id {
+            AlgebraicExpression::Number(id) => Some(*id),
+            _ => None,
+        }
+    }
+}
+
+impl<T: FieldElement> BusReceive<T> {
+    /// Returns whether the receive has an unconstrained multiplicity.
+    /// If it does not, this typically means that the multiplicity is binary constraint.
+    /// Unconstrained receives are equivalent to the RHS of a lookup;
+    /// binary-constrained receives are equivalent to the RHS of a permutation.
+    pub fn is_unconstrained(&self) -> bool {
+        // TODO: This always works in practice, but we should properly check the
```

Comment:
Which also means that this interface won't work if we do it properly :(

Reply:
We do have all identities when doing the conversion, so we could at least detect bit constraints and store it in the bus receive. In all cases I'm aware of, it would either be constrained to binary or unconstrained.

Reply:
I know I just meant that `fn is_unconstrained(&self)` doesn't work like that.